Edit options when initialized, Fix cleanup when recompiling

rogerbarton opened the pull request:
1. Allows editing options when initialized, this is done by having a separate copy of the Options in the script as in the manipulator. Changes are applied when entering play mode or can be applied immediately. This also avoids bugs when using enableInEditMode and changing the options when already initialized.
2. Properly cleans up when recompiling, the problem is that OnDestroy is not called when recompiling.

See commit messages for details.

rogerbarton committed:
Allow editing options when initialized, DllManip has a separate Options copy to DllManipScript

When DllManipulator is initialized we pass a copy of the current DllManipulatorScript options. This allows us to keep editing the options when the manipulator is initialized/dlls are loaded as we are editing a copy. This is particularly useful when using enableInEditMode.

Changes are always applied at OnEnable as before. To allow modifying without play/stop a DllManipulatorScript.Reinitialize() has been added which can be called via GUI (see DetectOptionChanges) which unloads dlls and initializes with new options. GUI is only shown when there are changes.

rogerbarton committed:
Fix OnDestroy not being called when recompiling

When recompiling only `OnDisable` is called, not `OnDestroy`. Now gets the callback for when the assembly is being reloaded, just when finished with compiling. This is called before `OnDisable`, so I just set a flag `_isRecompiling`.

Previously the native function `UnityPluginUnload` was not being called causing various bugs.

rogerbarton commented:
![PR20-gui-preview](https://user-images.githubusercontent.com/42413282/80821688-842e1c80-8bd9-11ea-898b-66ccbb3a0e90.gif)

mcpiroman commented on scripts/Editor/DllManipulatorEditor.cs:
```diff
@@ -210,8 +266,6 @@ private void DrawOptions(DllManipulatorOptions options)
         {
             var guiEnabledStack = new Stack<bool>();
             guiEnabledStack.Push(GUI.enabled);
```
If you remove this you could also remove the stack thing, this is the push here and corresponding pop below.

mcpiroman commented on scripts/Editor/DllManipulatorEditor.cs:
```diff
@@ -15,7 +15,7 @@ namespace UnityNativeTool.Internal
     {
         private static readonly string INFO_BOX_GUI_CONTENT = 
-            "Mocks native functions to allow manually un/loading native DLLs. DLLs are always unloaded at OnDestroy.";
+            "Mocks native functions to allow manually un/loading native DLLs. DLLs are always unloaded at OnDestroy. Changes below are always applied at OnEnable.";
```
```suggestion
            "Mocks native functions to allow manually un/loading native DLLs. DLLs are always unloaded at OnDestroy. Configuration changes below are always applied at OnEnable.";
```

mcpiroman commented on scripts/Editor/DllManipulatorEditor.cs:
```diff
@@ -93,12 +106,67 @@ public override void OnInspectorGUI()
+            
+            // When enabling enableInEditMode for the first time, allow immediately initializing without waiting for OnEnable
+            if(DllManipulator.Options == null && t.Options.enableInEditMode && !EditorApplication.isPlaying && 
+               GUILayout.Button(INITIALIZE_ENABLED_EDIT_MODE_GUI_CONTENT))
```
I would maybe move that button to be close to the 'enable in edit mode' option

mcpiroman replied:
I mean you enable it and this shows up right next to it so that you know you have to click it first for it to work, sth like warning button i guess

rogerbarton replied:
I think its reasonably clear when the button pops up. If the user immediately clicks the initialize button and then changes the options below they will have to unnecessarily reinitialize. Having the button at the bottom implies that the other options also affect its behaviour.

Also in the code it is clearer as the functionality about re/initializing is together in `DetectOptionChanges`.

mcpiroman replied:
Yup

mcpiroman replied:
Although, it looks like `Options` in `DllManipulator` can now be private?

rogerbarton replied:
We use it in the Script and Editor to check if the DllManip is initialized/there are changes to be applied. We could make the setter private?

mcpiroman replied:
yes

mcpiroman commented on scripts/DllManipulatorScript.cs:
```diff
@@ -110,10 +130,31 @@ public static void InvokeMainThreadQueue()
+                if (_isRecompiling)
+                {
+                    _isRecompiling = false;
+                    OnDestroy();
```
I don't really like calling a callback method so I'd be after creating separate one

rogerbarton replied:
Created a Reset()

mcpiroman commented on scripts/DllManipulatorScript.cs:
```diff
@@ -84,12 +87,29 @@ private void Initialize()
+            if (DllManipulator.Options != null)
+                DllManipulator.Reset();
+            
+#if UNITY_EDITOR
```
I think if this is not going to actually perform the initialization again, then this method shouldn't be called in the first place (it doesn't actually reinitialize).

rogerbarton replied:
Yes, I've changed it although this requires that the Initialize function is public.

mcpiroman commented on scripts/Editor/DllManipulatorEditor.cs:
```diff
@@ -93,12 +106,67 @@ public override void OnInspectorGUI()
+                        t.Reinitialize();
+                }
+                else if(GUILayout.Button(REINITIALIZE_WITH_CHANGES_LAZY_GUI_CONTENT))
+                    t.Reinitialize();
```
nit: parentheses there

mcpiroman commented on scripts/DllManipulatorScript.cs:
```diff
@@ -84,12 +87,29 @@ private void Initialize()
+            if(_singletonInstance != this)
+                return;
+            
+            if (DllManipulator.Options != null)
```
I'm not certain, but it appears to me that we could reduce the number of different places when options are stored and move them from this class to the `DllManipulator` itself? TBH I don't really like how it's scattered.

rogerbarton replied:
Yes its a bit scattered but I think this makes sense:
1. DllManipulatorScript has an Options copy which are seen in the inspector.
1. DllManipulator has Options which are currently used.
1. DllManipulatorEditor has Options solely for detecting changes in the inspector.

Each .cs only has to consider one Options (except the editor). Having the previousInspectorOptions from the editor in DM means DM has a copy which it never uses. I would leave it as is personally.

mcpiroman replied:
ya, sure

mcpiroman reviewed (commented):
Some quick notes I left here, will have a closer look soon, lgtm though, thanks!

rogerbarton committed:
Update scripts/Editor/DllManipulatorEditor.cs

Co-authored-by: mcpiroman <38111589+mcpiroman@users.noreply.github.com>

rogerbarton committed:
Review changes

Also removed check if singletonInstance in Reinitialize as this should never occur

mcpiroman commented on scripts/DllManipulatorScript.cs:
```diff
@@ -161,15 +157,18 @@ private void OnDisable()
-                    DllManipulator.Reset();
-                _singletonInstance = null;
-            }
+        public void Reset()
```
This one doesn't need to be public right?

rogerbarton replied:
No, its now private

mcpiroman commented on scripts/DllManipulator.cs:
```diff
@@ -123,6 +126,8 @@ public static void Reset()
             _customAfterUnloadTriggers?.Clear();
             _customBeforeUnloadTriggers?.Clear();
+
+            Options = null;
```
Wait a sec, what does it do?

rogerbarton replied:
We delete the DllManip Options copy. This is also used as an indicator to whether the DllManip is initialized.

mcpiroman replied:
I mean, how does it consume the options when they are null? Sry if I miss sth

mcpiroman replied:
Ah nvm, on the diff it blended in as though it was called in Initalize, not Reset, sorry

rogerbarton committed:
Small fix

rogerbarton committed:
DllManip.Options private set

mcpiroman commented:
LGTM
